### PR TITLE
Fy bmatrix perturbation and perturbation

### PR DESCRIPTION
**Pull request by AitorBengoechea**

### CODE REVIEW

File: sandy/fy.py
Comment (luca-fiorito-11) [inline sandy/fy.py:276]: you cannot test with a B-matrix for only 3 nuclides

File: sandy/core/endf6.py
Comment (luca-fiorito-11) [inline sandy/core/endf6.py:90]: `list` or `'all'`

File: sandy/core/endf6.py
Comment (luca-fiorito-11) [inline sandy/core/endf6.py:91]: `list`

File: sandy/core/endf6.py
Comment (luca-fiorito-11) [inline sandy/core/endf6.py:170]: PEP8

File: sandy/core/endf6.py
Comment (luca-fiorito-11) [inline sandy/core/endf6.py:273]: use `isinstance`.
I don't like this...what if instead of a list you get an array, or a tuple...it will break!

Check if `hasattr(zam, "__len__")` to see if it is an iterable

File: sandy/core/endf6.py
Comment (luca-fiorito-11) [inline sandy/core/endf6.py:279]: I prefer:
`tape = tape.merge(foo_get(filename, url))`

**Review by luca-fiorito-11 — CHANGES_REQUESTED** [review]

File: sandy/fy.py
Comment (AitorBengoechea) [reply sandy/fy.py:245]: I have changed all the PEP 8 bugs, except for the too long line bugs. 
When I do numpy.setdiff1d(ar1, ar2, assume_unique=False), the outpur is: 1D array of values in ar1 that are not in ar2. So, the only value I have to return is the perturbed value. Test:
>>>a = np.array([[1,7,3],[9,5,6]])
>>>b = np.array([[1,2,3],[4,5,6]])
>>>np.setdiff1d(a, b)
array([7, 9])

File: sandy/fy.py
Comment (AitorBengoechea) [reply sandy/fy.py:276]: I have changed the whole test, so that the IFYs of the 3 cores change to make it a valid test.

File: sandy/core/endf6.py
Comment (AitorBengoechea) [reply sandy/core/endf6.py:273]: I thought that with isinstance will not break because i introduce more ways to group data

**Comment by AitorBengoechea** [comment]
@luca-fiorito-11 I update the version with the changes review that you did.

**Comment by luca-fiorito-11** [comment]
@EnricaBelfiore , can you have a look?

File: sandy/core/endf6.py
Comment (luca-fiorito-11) [inline sandy/core/endf6.py:18]: we usually import it as 
```
import numpy as np
```

File: sandy/core/endf6.py
Comment (luca-fiorito-11) [inline sandy/core/endf6.py:93]: And what about a pandas.Series, or a Map object?
You should include all possible types of iterable...then the only way is to check if they have dunder `__len__`

File: sandy/core/endf6.py
Comment (luca-fiorito-11) [inline sandy/core/endf6.py:275]: rewrite this according to the previous comment

File: sandy/core/endf6.py
Comment (luca-fiorito-11) [inline sandy/core/endf6.py:281]: a simple else should be ok...what if ZAM is a float? e.g., 922350.0

File: sandy/fy.py
Comment (luca-fiorito-11) [inline sandy/fy.py:283]: I still believe you cannot use the B-matrix this way....I mean, in this test

File: sandy/fy.py
Comment (luca-fiorito-11) [inline sandy/fy.py:276]: You are missing the note we said we had to add.

```
Notes
-----
.. note:: write your note
```

File: sandy/fy.py
Comment (luca-fiorito-11) [inline sandy/fy.py:307]: CFY = Q * IFY

File: sandy/fy.py
Comment (luca-fiorito-11) [inline sandy/fy.py:328]: same problem as for previous method

File: sandy/fy.py
Comment (luca-fiorito-11) [inline sandy/fy.py:345]: there is no need for renaming

Comment (luca-fiorito-11) [reply sandy/fy.py:345]: Do you need to convert it to frame?

**Review by luca-fiorito-11 — CHANGES_REQUESTED** [review]

File: sandy/fy.py
Comment (AitorBengoechea) [reply sandy/fy.py:283]: The aim of the test is: 
Select related cores by decay (so that the B matrix is not zero) and create a variable (minimal_fytest_2) with these cores and invented IFY and CFY values. 
As IFY and CFY are invented, when applying bmatrix, all IFY values will change, since IFY and CFY are not related. In this way, by manually performing the multiplication between the matrix (1-B) and the vector CFY, we can check that the method results are correct.  
Finally, we compare the values of the original FY object and the one to which we have applied bmatrix. The only different values have to be the IFY values.

**Comment by AitorBengoechea** [comment]
Luca do you see the changes I have made?
I'm working on my personal computer and I'm not sure if you can see the changes I've made. In my Github account if I log into the branch I am able to see them, but I'm not sure if in the pull request you can see them.

**Comment by luca-fiorito-11** [comment]
@AitorBengoechea , yes I can see them!

**Review by EnricaBelfiore — APPROVED** [review]

File: sandy/core/endf6.py
Comment (luca-fiorito-11) [inline sandy/core/endf6.py:276]: @EnricaBelfiore, you shouldn't approve this! This will crash if zam is an integer.
@AitorBengoechea , I am surprised you don't get errors...probably because you don't have any test where you ask for a single decay nuclide (only all or list). This test should be added, otherwise these mistakes will not be detected.

File: sandy/core/endf6.py
Comment (luca-fiorito-11) [inline sandy/core/endf6.py:274]: I expected something like this example:

inputs
-------
```
url = sandy.libraries.URL_DECAY_JEFF_33_IAEA
files = sandy.libraries.DECAY_FILES_JEFF_33_IAEA
zam = (551340, 551330, 601480)
foo_get = sandy.Endf6.from_zipurl
```
what I would do
---------------
```
if hasattr(zam, "__len__"):
    tapes = map(lambda x: foo_get(files[x], url), zam)
    tape = reduce(lambda x, y: x.add_sections(y.data), tapes)
else:
    tape = foo_get(files[zam], url)
```
this should work both for scalar or iterables...lists, tuples, arrays, series, map objects, filter objects and so on.

It is taken from `Endf6.reduce`, as i mentioned to you!
Please check it before implementing it!

**Review by luca-fiorito-11 — CHANGES_REQUESTED** [review]

File: sandy/core/endf6.py
Comment (AitorBengoechea) [reply sandy/core/endf6.py]: The tests for individual nucleus are put inside the get_endf6_file function, but even passing pytest with the mentioned error, pytest still pass the tests.
The program breaks in __FormattedFile.merge, i do not know why

File: sandy/core/endf6.py
Comment (luca-fiorito-11) [inline sandy/core/endf6.py:274]: Last comment, this section is better like this, right?!

Yours might fail in some specific case!

```
    if hasattr(zam, "__len__"):
        tapes = map(lambda x: foo_get(files[x], url), zam)
        tape = reduce(lambda x, y: x.add_sections(y.data), tapes)
    else:
        if str(zam).lower() == 'all':
            if kind.lower() == 'xs':
                raise ValueError("'all' option is not available for xs")
            text = "".join([foo_read(name, url) for name in files.values()])
            tape = Endf6.from_text(text)
        else:
            tape = foo_get(files[zam], url)
```

File: sandy/core/endf6.py
Comment (luca-fiorito-11) [reply sandy/core/endf6.py:18]: why I don't see `import numpy as np`?

Anyway, do not import numpy since you don't use it

File: sandy/core/endf6.py
Comment (luca-fiorito-11) [reply sandy/core/endf6.py:274]: can you check this?

**Review by luca-fiorito-11 — CHANGES_REQUESTED** [review]

File: sandy/core/endf6.py
Comment (AitorBengoechea) [reply sandy/core/endf6.py:274]: In this order we have a problem: str variables has the atributte __len__

Comment (AitorBengoechea) [reply sandy/core/endf6.py:274]: So the program fail

Comment (luca-fiorito-11) [reply sandy/core/endf6.py:273]: ok, then keep it as it is!

**Review by luca-fiorito-11 — APPROVED** [review]

**Pull request merged**
